include/gul14/ThreadPool.h

```diff
@@ -0,0 +1,578 @@
+ * stored in a queue and executed in the order they were added. This makes a ThreadPool
+ * with a single thread effectively a serial task queue.
+ *
+ * Each task is associated with a TaskHandle. This handle is returned by add_task() and
+ * can be used to query the status of the task or to remove it from the queue via
+ * cancel(). Tasks that are already running cannot be canceled.
+ *
+ * Tasks can be scheduled to start at a specific time point or after a certain delay. Each
+ * task can also be given a name, which is mainly useful for debugging.
+ *
+ * \ref thread_pool.cpp "thread_pool.cpp".
+ *
+ * \example thread_pool.cc
+ * An example on how to use the ThreadPool class to schedule tasks.
+ *
+ * @{
+ */
+
+/**
+ * A handle for a task that has (or had) been enqueued on a ThreadPool.
+ *
+ * \code{.cpp}
+ * ThreadPool pool(1);
+ * auto task = pool.add_task([]() { return 42; });
+ * while (not task.is_complete())
+ * {
+ *     std::cout << "Waiting for task to complete...\n";
+ *     sleep(0.1);
+ * }
+ * std::cout << "Task result: " << task.get_result() << "\n";
+ * \endcode
+ */
+template <typename T>
+class TaskHandle
+{
+public:
+    /**
+     * Remove the task from the queue if it is still pending.
+     *
+     * This call has no effect if the task is already running.
+     *
+     * \returns true if the task was removed from the queue, false if it was not found in
+     *          the queue (e.g. because it is already running).
+     *
+     * \exception std::logic_error is thrown if the associated thread pool does not
+     *            exist anymore.
+     */
+    bool cancel() const { return detail::cancel_task(pool_, id_); }
+
+    /// Return the ID of the task.
+    TaskId get_id() const noexcept { return id_; }
+
+    /**
+     * Block until the task has finished and return its result.
+     *
+     * If the task finished by throwing an exception, get_result() rethrows this
+     * exception.
+     */
+    T get_result() { return future_.get(); }
+
+    /// Determine whether the task has finished.
+    bool is_complete() const { return future_.wait_for(std::chrono::seconds(0)) == std::future_status::ready; }
+
+    /**
+     * Return true if the task is still waiting to be started.
+     *
+     * \exception std::logic_error is thrown if the associated thread pool does not
+     *            exist anymore.
+     */
+    bool is_pending() const { return detail::is_pending(pool_, id_); }
+
+    /**
+     * Return true if the task is currently being executed.
+     * \exception std::logic_error is thrown if the associated thread pool does not
+     *            exist anymore.
+     */
+    bool is_running() const { return detail::is_running(pool_, id_); }
+
+private:
+    friend class ThreadPoolEngine;
+
+    std::future<T> future_;
+    TaskId id_;
+    std::weak_ptr<ThreadPoolEngine> pool_;
+
+    TaskHandle(TaskId id, std::future<T> future, std::shared_ptr<ThreadPoolEngine> pool)
```

Comment:
The constructor is private? :thinking: 
I'm not much of a fan of these 'security' features. If someone wants to 'mess around' with it, why try to hinder it?
Also not very good for tests ;)

Reply:
Well, just because it does not make sense is not a real reason to prevent someone doing it.
There is no real need for this, and if there is no need (like keep some invariant etc) trying to prevent users doing stuff can be very frustrating if there ever IS a case where someone needs to do something that we do not forsee at the moment and cant, because 'we need security' and 'we know better than the user, she shall not construct a TaskHandle on her own'.

Reply:
The `TaskHandle` constructor is now public. The class is now nested inside `ThreadPool`, and I reckon the chances of someone accidentally trying to instantiate a `ThreadPool::TaskHandle` are slim.